packages/webapp/src/components/Icons/icons.tsx

```diff
@@ -63,8 +63,29 @@ import { ReactComponent as SoilAmendmentTask } from '../../assets/images/task/So
 import { ReactComponent as MoreHorizontalIcon } from '../../assets/images/more-horizontal.svg';
 import { ReactComponent as PlusCircleIcon } from '../../assets/images/plus-circle.svg';
 import { ReactComponent as TrashIcon } from '../../assets/images/animals/trash_icon_new.svg';
+import { FunctionComponent } from 'react';
 
-const iconMap = {
+const animalIconMap = {
+  CATTLE: CattleIcon,
+  CHICKEN: ChickenIcon,
+  PIGS: PigIcon,
+  BATCH: BatchIcon,
+  BATCH_GREEN: BatchIconGreen, // svgColorFill does not work with this icon
+  CUSTOM_ANIMAL: CustomAnimalIcon,
+  ALPACA: AlpacaIcon,
+  GOAT: GoatIcon,
+  RABBIT: RabbitIcon,
+  SHEEP: SheepIcon,
+};
+
+const AnimalTypeIcons = Object.keys(animalIconMap);
+export type AnimalTypeIconKey = (typeof AnimalTypeIcons)[number];
```

Comment:
I think I personally preferred the original way (keeping a separate string array for the animal icon names in particular, since they are the ones with a type guard).  And as I said, you can enforce that they all exist in `iconMap` by typing them as `IconName[]`.

But if you prefer this way to avoid the redundancy, it should work, but I don't think you can type off of the `Object.keys()` -- it will just be a string array. Typing off of the map should be okay, though:

```ts
export type AnimalTypeIconKey = keyof typeof animalIconMap;
```

Reply:
I tried enforcing it with `IconName[]` but it is a circular reference -- its slightly tricky but it is technically already enforced by `AnimalTypeIconKey` being used in the definition for `IconName`.

Decided to revert the change and keep the strings. It seems easier to convert an array to a type than the opposite.

Reply:
Sorry to not be more clear. This is what I was originally suggesting:

```ts
const AnimalTypeIcons: IconName[] = [ // <-- type as IconName[] so you can't add just any string
  'CATTLE',
  'CHICKEN',
  'PIGS',
  'BATCH',
  'BATCH_GREEN',
  'CUSTOM_ANIMAL',
  'ALPACA',
  'GOAT',
  'RABBIT',
  'SHEEP',
] as const;

export type AnimalTypeIconKey = (typeof AnimalTypeIcons)[number];

export const isAnimalTypeIconKey = (iconKey: string): iconKey is AnimalTypeIconKey => {
  return AnimalTypeIcons.includes(iconKey as AnimalTypeIconKey);
};

export type IconName = keyof typeof iconMap; // <-- this is my main suggestion

export const iconMap = {  // <-- remove the type here to avoid circular reference
  // ...
```

The point is to let the `IconName` type be automatically generated from the keys of iconMap, and you would not type iconMap in that case to avoid the circular reference.

However, your second method (creating an `animalIconMap` and spreading it into `iconMap`) was also totally fine -- you didn't have to revert it! It's nice because it has no redundancies and let's you type iconMap -- maybe it's the best of the three. But in any case, as I said in my comment, you just needed to change a single line (the type export) and it would behave as before too.

```ts
const animalIconMap = {
  CATTLE: CattleIcon,
  CHICKEN: ChickenIcon,
  PIGS: PigIcon,
  BATCH: BatchIcon,
  BATCH_GREEN: BatchIconGreen,
  CUSTOM_ANIMAL: CustomAnimalIcon,
  ALPACA: AlpacaIcon,
  GOAT: GoatIcon,
  RABBIT: RabbitIcon,
  SHEEP: SheepIcon,
};

const AnimalTypeIcons = Object.keys(animalIconMap);
export type AnimalTypeIconKey = keyof typeof animalIconMap; <-- only change needed

// TypeGuard for animal type icons
export const isAnimalTypeIconKey = (iconKey: string): iconKey is AnimalTypeIconKey => {
  return AnimalTypeIcons.includes(iconKey); <-- bonus: type casting not necessary this way
};

export const iconMap: Record<string, FunctionComponent> = { <-- another bonus: can still keep this type
  // ...
```

Either of those solutions will work without manually listing out the strings in `IconName`.

Reply:
So I tried that but the following happens: 

![Screenshot 2024-09-03 at 6 39 07 PM](https://github.com/user-attachments/assets/3a82c774-fd9e-4664-b0ea-75113626c78f)

IconName needs to be specific.

Reply:
I think then removing the `Record` typing fixes that issue -- but then iconMap isn't typed haha

Reply:
Ahh your right, sorry -- in both cases you can't retain `Record<string, FunctionComponent>` without that breaking `keyof typeof iconMap` 🤔 That seems strange -- there has to be a way! 

I think I still prefer `iconMap` untyped rather than the behemoth of manual types (that will keep having to grow), but maybe Anto can rescue us with a solution that's best of both worlds 😂

Reply:
@kathyavini I like this solution actually! 

I was looking for more info about it and the terminology I found is "Constrained Identity Function": https://kentcdodds.com/blog/how-to-write-a-constrained-identity-function-in-typescript

There might also be an option to use `satisfies` if our ts is up to date!